### ISSUE TITLE
Python wish: forgot to pass value for kwarg

### ISSUE THREAD

**Issue by jnothman** [lgtmhq/lgtm-queries#14]
A kind of bug we have seen repeatedly in scikit-learn has a form roughly like this:

```python
def _helper(x, y, fancy_name=5):
    do something


def user_facing(x, y, fancy_name=5):
    _helper(x, y)
```

Here, `user_facing` fails to pass `fancy_name` to `_helper` which has a parameter of the same name.

Sometimes it's more complicated as it looks like:

```python
class UserFacing:
    def __init__(self, fancy_name=5):
        self.fancy_name = 5
    def some_func(self, x, y):
        _helper(x, y)
```

This is the sort of thing that should be caught by proper integration tests, or by randomised testing that ensures that the outcome can change as a result of the parameter being changed; but often it is missed.

I don't know what are the best indicators that this kind of bug exists or is a false positive:
* match of names between local/class context and called function's kwarg
* a requirement that the variable name be non-trivial
* the variable otherwise being unused (by some definition)
* the common case that other kwargs being passed to the helper match, but this one (or two) is absent
* that the kwarg was added to the caller and the callee in the same commit

I certainly don't know where to start expressing this kind of query with QL. But I thought I would enjoy challenging LGTM to the task. :)

Ping @amueller, @lesteve, or others, do you agree this is a common issue? How would you characterise it?

**Comment by lesteve** [comment]
Out of curiosity do you have a scikit-learn issue related to this kind of thing?

**Comment by jnothman** [comment]
are you asking me whether i have made an issue at scikit-learn to test for
such things through random parameters? No, though I have thought to do so.
I don't think it's something we'd want in the pull request CI as it would
presumably be slow. It also requires specification of valid parameter
ranges (and which parameters, such as n_jobs may not affect the model), and
assert_models_not_equal. The former is something the auto ml packages may
have but scikit-learn doesn't, and the latter is languishing somewhere in a
pull request. So we're not up to doing it by traditional testing.

 But it is hard to spot by eye, *particularly by looking at a diff*. Hence
the sort of analysis offered here.

**Comment by lesteve** [comment]
> are you asking me whether i have made an issue at scikit-learn to test for such things through random parameters?

Sorry for not being clear. You were saying:

> A kind of bug we have seen repeatedly in scikit-learn has a form roughly like this:

I was just wondering whether you could point me to an already existing scikit-learn issue where this kind of bug showed up.

**Comment by jnothman** [comment]
I might well be exaggerating the simplicity of the case. It's hard to find them in hindsight, but I'm sure I've seen it in more bugs than this, as well as in reviewing.

* https://github.com/scikit-learn/scikit-learn/issues/7913 is the sort of thing. Again here note that there is parallelism between the if and the else cases; normalize is passed in the if but not the else case.
* https://github.com/scikit-learn/scikit-learn/issues/1178 reports a case for `n_jobs` not being passed; the fix includes `random_state` as well.
* https://github.com/scikit-learn/scikit-learn/issues/9032 fixes a case where `n_iter` was not passed but there the param names to not match exactly.

**Comment by jnothman** [comment]
* https://github.com/scikit-learn/scikit-learn/issues/8173
* https://github.com/scikit-learn/scikit-learn/commit/07cede7 mentions a case where an underlying estimator did not receive a `sample_weight` from a meta-estimator, but this may be a bit far
* https://github.com/scikit-learn/scikit-learn/pull/5858/files#diff-2229647ab9a84dc25ab3d3d13800cb43R1790 is a similar thing, although there a fixed value was mistakenly passed instead of relying on the default
* https://github.com/scikit-learn/scikit-learn/pull/2617 similarly
* https://github.com/scikit-learn/scikit-learn/commit/f157606 shows this kind of thing being fixed while devloping a PR
* https://github.com/scikit-learn/scikit-learn/commit/6eee7a5
* https://github.com/scikit-learn/scikit-learn/pull/4654

**Comment by lesteve** [comment]
Thanks for the details @jnothman it does seem like a recurring thing indeed!

**Comment by ogrisel** [comment]
+1 as well.

**Comment by amueller** [comment]
Yeah definitely a recurring thing. I think a function argument that is not used would maybe be the easiest way to start. I think flake8 should be able to find that, but it doesn't :-/

```python
def func():
    x = 5
    return 6
```
raises a flake8 error, but

```python
def func(x=5):
    return 6
```
does not.

Maybe because that is actually intended behavior in many cases? But it would still be good to be able to detect it.

**Comment by jnothman** [comment]
unused function args might be good for a once off analysis but is far from
sufficient for an alert. Inheritance and API consistency means arguments
are often not used.

**Comment by markshannon** [comment]
I've written a simple query that handles the first case: a function calling another function, both of which have the same parameter, but the value is not passed in the call.
Here is the query and the result of running it on scikit-learn:
https://lgtm.com/query/2014811136/project:1870011/lang:python/

I don't know scikit-learn at all, but some of the results look like they might be the sort of thing you are looking for. 
https://lgtm.com/projects/g/scikit-learn/scikit-learn/snapshot/543f3dc2eb58ad809244b94a1dc5f38301f124b1/files/sklearn/neighbors/tests/test_neighbors.py#L965
(`n_nieghbours` is missing)

https://lgtm.com/projects/g/scikit-learn/scikit-learn/snapshot/543f3dc2eb58ad809244b94a1dc5f38301f124b1/files/sklearn/calibration.py#L190
(`sample_weight`is missing)

**Comment by jnothman** [comment]
Awesome! I've not taken a look yet, but it would be great to see when this
kind of error gets added to the code.

On 8 Jul 2017 1:35 am, "Mark Shannon" <notifications@github.com> wrote:

I've written a simple query that handles the first case: a function calling
another function, both of which have the same parameter, but the value is
not passed in the call.
Here is the query and the result of running it on scikit-learn:
https://lgtm.com/query/2014811136/project:1870011/lang:python/

I don't know scikit-learn at all, but some of the results look like they
might be the sort of thing you are looking for.
https://lgtm.com/projects/g/scikit-learn/scikit-learn/snapshot/
543f3dc2eb58ad809244b94a1dc5f38301f124b1/files/sklearn/neighbors/tests/test_
neighbors.py#L965
(n_nieghbours is missing)

https://lgtm.com/projects/g/scikit-learn/scikit-learn/snapshot/
543f3dc2eb58ad809244b94a1dc5f38301f124b1/files/sklearn/calibration.py#L190
(sample_weightis missing)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/lgtmhq/lgtm-queries/issues/14#issuecomment-313715838>,
or mute the thread

**Comment by jnothman** [comment]
Btw #9299 is another case with class parameters

**Comment by jnothman** [comment]
I've taken a look at your query, and it's nice to see how simple this is to express.

Neither of the two examples cited are true positives. The n_neighbors case is a special one where there is a method parameter falling back to a class parameter. It's a fairly unusual case in our API.

However, the first two alerts I've examined myself are highly suspicious, so this is definitely a good start!

I'll think about reducing false positives

**Comment by jnothman** [comment]
@markshannon how long should I expect the query to run if I modify it? It's been running for two hours. I presume that is not the timeframe in which LGTM runs when used as CI.

**Comment by jnothman** [comment]
I suspect many of these can be picked up by parameter never used, as @amueller suggests, but we would need to work out how to make exceptions for them to reduce FPs.

**Comment by jnothman** [comment]
I would like to add to the alert message "and is unused in the caller" and "and is unused in the callee" as appropriate. I'll have to work that out.

**Comment by sjvs** [comment]
@jnothman: @markshannon is currently attending EuroPython, so he may be a bit slow to reply to your question. I know that there he accidentally posted a different link in his original comment to this issue; can you please verify that you modified the right query? This is the correct URL: https://lgtm.com/query/2028770042/project:1870011/lang:python/

The running time of a query very heavily depends on a number of factors, including the size of the code base, the complexity of the query, and how much we succeed in optimising it in our engine. Our standard analysis on scikit-learn (with the default query pack, containing 140 Python queries) takes just over half an hour on our cloud infrastructure. If you're interested in the details, you can view the log files here: https://lgtm.com/projects/g/scikit-learn/scikit-learn/logs/rev/a6753f3ed38d25cec0af8ed95697f48eaacaed24 (make sure to click through to the 'Analysis' tab)

We're continuously working on the efficiency of our optimiser and query engines to make things run faster. Later this year we'll make it possible for people to write/amend their own queries and include results from those queries in the PR/CI integration (as well as the standard visualisations shown on lgtm.com).

Thanks again for your feedback — great to hear from early adopters :)

**Comment by jnothman** [comment]
I've looked through these alerts (ignoring test files and others) and raised the relevant issues, all back-referenced here.

In many cases, there is a conditional expression where the `if` uses the parameter and the `else` does not, or vice-versa. Sometimes this is intentional, particularly when the condition incorporates the parameter, and the alert is a false positive. Often the conditioning decides between two algorithmic strategies and we should probably be passing the parameter in both cases.

While I'm not sure it would be useful to track this kind of alert overall (too many false positives), tracking newly created alerts of this kind (as an optionally relevant code smell) would be worthwhile. Does lgtm.com (hope to) distinguish between essential alert queries and smell type alert queries where false positives are no big deal?

**Comment by sjvs** [comment]
@jnothman: thanks for referring back to this and filing the issues. Great to see that our results are helpful.

Regarding distinguishing between different types of alerts: as of yesterday the list of alerts are displayed in a different way, which distinguishes between different alert severities. We're looking to implement support for more flexible query configurations, e.g. include custom queries in analyses and functionality to run a different set of queries for pull requests reviews. That could be used for this purpose.

Thanks again for your suggestions — if you have any more, let me know!
